api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/txml/StartedArtOnLastClinicalContactCalculation.java

```diff
@@ -0,0 +1,126 @@
+import org.openmrs.module.reporting.data.patient.definition.SqlPatientDataDefinition;
+import org.openmrs.module.reporting.evaluation.parameter.Parameter;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class StartedArtOnLastClinicalContactCalculation extends AbstractPatientCalculation {
+
+  @Autowired private HivMetadata hivMetadata;
+
+  private static final String ON_OR_BEFORE = "endDate";
+
+  @Override
+  public CalculationResultMap evaluate(
+      Collection<Integer> cohort,
+      Map<String, Object> parameterValues,
+      PatientCalculationContext context) {
+    CalculationResultMap lastClinicalContactMap = getLastClinicalContactMap(cohort, context);
+
+    Boolean lessThan90Days = (Boolean) parameterValues.get("lessThan90Days");
+    Date endDate = (Date) context.getFromCache("onOrBefore");
+
+    CalculationResultMap lessThan90DaysMap = new CalculationResultMap();
+    CalculationResultMap moreThan90DaysMap = new CalculationResultMap();
+    CalculationResultMap artStartDates =
+        calculate(
+            Context.getRegisteredComponents(InitialArtStartDateCalculation.class).get(0),
+            cohort,
+            parameterValues,
+            context);
+    if (endDate == null) {
+      endDate = (Date) context.getFromCache("onOrBefore");
+    }
+    if (endDate != null) {
+      for (Integer patientId : cohort) {
+        Date artStartDate =
+            InitialArtStartDateCalculation.getArtStartDate(patientId, artStartDates);
+        Date lastClinicalContact =
+            EptsCalculationUtils.resultForPatient(lastClinicalContactMap, patientId);
+
+        if (artStartDate != null && lastClinicalContact != null) {
+          Date onArtForLessThan90Days = EptsCalculationUtils.addDays(artStartDate, 90);
+
+          if (onArtForLessThan90Days.compareTo(lastClinicalContact) < 0) {
+            lessThan90DaysMap.put(patientId, new BooleanResult(true, this));
+          }
+          if (onArtForLessThan90Days.compareTo(lastClinicalContact) >= 0) {
+            moreThan90DaysMap.put(patientId, new BooleanResult(true, this));
+          }
+        }
+      }
+      if (lessThan90Days) {
+        return lessThan90DaysMap;
+      } else {
+        return moreThan90DaysMap;
+      }
+    } else {
+      throw new IllegalArgumentException(String.format("Parameter %s must be set", ON_OR_BEFORE));
+    }
+  }
+
+  /**
+   * @param cohort
+   * @param context
+   * @return
+   */
+  private CalculationResultMap getLastClinicalContactMap(
+      Collection<Integer> cohort, PatientCalculationContext context) {
+    SqlPatientDataDefinition sqlPatientDataDefinition = new SqlPatientDataDefinition();
+    sqlPatientDataDefinition.addParameter(new Parameter("onOrBefore", "onOrBefore", Date.class));
+    sqlPatientDataDefinition.addParameter(new Parameter("location", "location", Location.class));
+
+    String query =
```

Comment:
why  you are  putting  pure query in  the Calculation, it would be better if  we use EptsCalculationUtils

Reply:
We were embedding a query into a calculation. Got a recommendation from the team @nthfloor  and @ynurmahomed